# Setup the training environment

Check if a Nvidia GPU is availiable.

If you get `/bin/bash: line 1: nvidia-smi: command not found`, select the
**T4 GPU** from withing the **Change runtime type** menu option.
Executing the training on a CPU takes ages! 🐌

In [ ]:
!nvidia-smi

Wed May 21 13:04:56 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   42C    P8              9W /   70W |       0MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

Install dependencies for the training process:

In [ ]:
!pip install pyyaml torch torchvision ultralytics

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 119.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 96.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 49.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 1.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 7.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 12.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 7.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 5.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 99.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 56.0 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
    Found existing installation: nvidia-nvjitlink-cu12 12.5.82
    Uninstalling

Set some variables:

In [ ]:
import os

DRIVE = os.path.join('./drive', 'MyDrive') # The path to your Google Drive (including mount point)
DATASET_ZIP = os.path.join(DRIVE, 'ml/var-ai/datasets/soccer-ball-1408.zip')

In [ ]:
DATASET_ZIP

'./drive/MyDrive/ml/var-ai/datasets/soccer-ball-1364.zip'

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


from google.colab import drive

mount_point = 'drive'
drive.mount(mount_point)

Create a folder structure for the dataset:

In [ ]:
!mkdir -p \
    /content/dataset/{images,labels}/{train,val} \
    /content/.tmp

----

# Prepare the Dataset


Clear the `.tmp`-folder containing an old unsplit dataset, e.g. when using a new/different dataset:

In [ ]:
!rm -r /content/.tmp/*

rm: cannot remove '/content/.tmp/*': No such file or directory


Extract dataset:

In [ ]:
import zipfile

with zipfile.ZipFile(DATASET_ZIP, 'r') as zip_ref:
    zip_ref.extractall('/content/.tmp')

Clear the `dataset`-directory, useful if you want to create a different split:

In [ ]:
!rm \
    /content/dataset/images/train/* \
    /content/dataset/images/val/* \
    /content/dataset/labels/train/* \
    /content/dataset/labels/val/*

Specify split ratio:

In [ ]:
SPLIT_RATIO = 0.8

Split dataset:

In [ ]:
import shutil
import random

from pathlib import Path

output_dir = 'dataset'
image_dir = Path('.tmp/images')
label_dir = Path('.tmp/labels')

image_files = list(image_dir.glob('*.jpg'))
random.shuffle(image_files)

split_index = int(len(image_files) * SPLIT_RATIO)
train_images = image_files[:split_index]
val_images = image_files[split_index:]

def move_files(image_list, subset):
    for image_path in image_list:
        label_path = label_dir / (image_path.stem + '.txt')

        img_dst = Path(output_dir) / 'images' / subset
        lbl_dst = Path(output_dir) / 'labels' / subset
        img_dst.mkdir(parents=True, exist_ok=True)
        lbl_dst.mkdir(parents=True, exist_ok=True)

        shutil.copy(image_path, img_dst / image_path.name)
        if label_path.exists():
            shutil.copy(label_path, lbl_dst / label_path.name)

move_files(train_images, 'train')
move_files(val_images, 'val')

Create a `yaml`-configuration file for the the dataset:

In [ ]:
import yaml

DATASET_CONFIGURATION_PATH = 'dataset.yaml'

dataset_configuration = {
    'path': '/content/dataset',
    'train': 'images/train',
    'val': 'images/val',
    'nc': 1,
    'names': ['sports ball']
}

with open(DATASET_CONFIGURATION_PATH, 'w') as config:
    yaml.dump(dataset_configuration, config)

----

# Model Training/Fine-Tuning

Our biggest most current [dataset](https://drive.google.com/file/d/1iboKyJKFsHlWk4dnXQ8VEUtO6kBDrZ_h/view?usp=drive_link) contains 1364 labeld images (as of 06.05.2025)

### Paremeters

Find an overview of all avaliable parameters [here](https://docs.ultralytics.com/modes/train/#train-settings).

Specify it's output path (preferably Google Drive so you don't lose training data when **timing out**):

In [ ]:
OUTPUT_DIR = os.path.join(DRIVE, 'ml/var-ai/models')
MODEL = 'yolo11s.pt'
TIME = 4.5
PATIENCE = 10
CACHE = True
SEED = 42
PLOTS = True
IMAGE_SIZE = 1280

### Learning Rate

Learning rate of `0.005` for 1-10k images

In [ ]:
LEARNING_RATE = 0.005

### Batch Size

The [batch size](https://docs.ultralytics.com/guides/model-training-tips/#how-to-train-a-machine-learning-model) specifies the number of data samples a machine learning model processes in a single training iteration. For example `16` means there are 16 images processed at a time before performing a weight update. This value should be set to the maximum a GPU/Processing-Unit can handle in order to **utilize** the **whole hardware**. For yolo11 there is `-1` which selects the optimal batch size based on the used hardware. Setting this too high results in **memory errors**.

One could use *trail-and-error* to find the optimal batch size by increasing the
the number until mentioned memory errors occure.

In [ ]:
BATCH_SIZE = -1

### Layer Freezing

For [transfer learning](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers/) (aka. fine tuning) a good tradeoff is `10` layers, this means the 10 most inner layers are kept, while `0` means the whole model is retrained. A higher number results in **faster training**, as the later layers are not retrained. A **drawback** of a higher freeze is a **lower percision**.

In [ ]:
FREEZE = 10

### Naming

For consistency purposes, choose a name that represents some important parameters e.g.:

In [ ]:
!ls -la /content/drive/MyDrive/ml/var-ai/models | cut -d ' ' -f 10 | grep 'yolo'

yolo11m-sports-ball-img1920-freeze10-crash
yolo11s-sports-ball-img1280-freeze10
yolo11s-sports-ball-img1920-freeze10
yolo11s-sports-ball-img640-freeze10


In [ ]:
import sys
sys.exit("Please enter a new model name:")

SystemExit: Please enter a new model name:

In [ ]:
NAME = 'yolo11s-sports-ball-img1280-freeze10-1408'

### Training the yolo model

In [ ]:
import torch
from ultralytics import YOLO

# Check if cuda is avaliable, if not, don't even start the training!
device = torch.device('cuda' if torch.cuda.is_available() else exit)

model = YOLO(MODEL)
model = model.to(device)
model.train(
    data=DATASET_CONFIGURATION_PATH,
    imgsz=IMAGE_SIZE,
    batch=BATCH_SIZE,
    device=device,
    freeze=FREEZE,
    lr0=LEARNING_RATE,
    project=OUTPUT_DIR,
    patience=PATIENCE,
    time=TIME,
    cache=CACHE,
    seed=SEED,
    plots=PLOTS,
    name=NAME)

Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/root/.config/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.


100%|██████████| 18.4M/18.4M [00:00<00:00, 270MB/s]


engine/trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=-1, bgr=0.0, box=7.5, cache=True, cfg=None, classes=None, close_mosaic=10, cls=0.5, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=dataset.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=100, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=10, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=1280, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.005, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolo11s.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=yolo11s-sports-ball-img1280-freeze10-1408, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto, overlap_mask=True, patience=10, perspective=0.0, plots=True, pose=12.0, pretrained=True, profile=False, project=./drive/MyDrive/ml/var-ai/models, 

100%|██████████| 755k/755k [00:00<00:00, 38.2MB/s]

Overriding model.yaml nc=80 with nc=1

                   from  n    params  module                                       arguments                     
  0                  -1  1       928  ultralytics.nn.modules.conv.Conv             [3, 32, 3, 2]                 
  1                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  2                  -1  1     26080  ultralytics.nn.modules.block.C3k2            [64, 128, 1, False, 0.25]     
  3                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
  4                  -1  1    103360  ultralytics.nn.modules.block.C3k2            [128, 256, 1, False, 0.25]    
  5                  -1  1    590336  ultralytics.nn.modules.conv.Conv             [256, 256, 3, 2]              
  6                  -1  1    346112  ultralytics.nn.modules.block.C3k2            [256, 256, 1, True]           
  7                  -1  1   1180672  ultralytics

 22                  -1  1   1511424  ultralytics.nn.modules.block.C3k2            [768, 512, 1, True]           
 23        [16, 19, 22]  1    819795  ultralytics.nn.modules.head.Detect           [1, [128, 256, 512]]          
YOLO11s summary: 181 layers, 9,428,179 parameters, 9,428,163 gradients, 21.5 GFLOPs

Transferred 493/499 items from pretrained weights
Freezing layer 'model.0.conv.weight'
Freezing layer 'model.0.bn.weight'
Freezing layer 'model.0.bn.bias'
Freezing layer 'model.1.conv.weight'
Freezing layer 'model.1.bn.weight'
Freezing layer 'model.1.bn.bias'
Freezing layer 'model.2.cv1.conv.weight'
Freezing layer 'model.2.cv1.bn.weight'
Freezing layer 'model.2.cv1.bn.bias'
Freezing layer 'model.2.cv2.conv.weight'
Freezing layer 'model.2.cv2.bn.weight'
Freezing layer 'model.2.cv2.bn.bias'
Freezing layer 'model.2.m.0.cv1.conv.weight'
Freezing layer 'model.2.m.0.cv1.bn.weight'
Freezing layer 'model.2.m.0.cv1.bn.bias'
Freezing layer 'model.2.m.0.cv2.conv.weight'
Freezing layer 'mod

100%|██████████| 5.35M/5.35M [00:00<00:00, 133MB/s]


AMP: checks passed ✅
train: Fast image access ✅ (ping: 0.0±0.0 ms, read: 2803.2±778.0 MB/s, size: 326.0 KB)


train: Scanning /content/dataset/labels/train... 1126 images, 0 backgrounds, 0 corrupt: 100%|██████████| 1126/1126 [00:00<00:00, 2081.77it/s]

train: New cache created: /content/dataset/labels/train.cache


WARNING ⚠️ cache='ram' may produce non-deterministic training results. Consider cache='disk' as a deterministic alternative if your disk space allows.


train: Caching images (2.9GB RAM): 100%|██████████| 1126/1126 [00:21<00:00, 53.00it/s]


albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))
AutoBatch: Computing optimal batch size for imgsz=1280 at 60.0% CUDA memory utilization.
AutoBatch: CUDA:0 (Tesla T4) 14.74G total, 0.18G reserved, 0.15G allocated, 14.42G free
      Params      GFLOPs  GPU_mem (GB)  forward (ms) backward (ms)                   input                  output
     9428179       86.19         1.242         91.28         550.2      (1, 3, 1280, 1280)                    list
     9428179       172.4         2.041         57.73         110.8      (2, 3, 1280, 1280)                    list
     9428179       344.8         3.274         98.08          98.9      (4, 3, 1280, 1280)                    list
     9428179       689.6         5.746         158.5         145.9      (8, 3, 1280, 1280)                    list
     9428179        1379        10

train: Scanning /content/dataset/labels/train.cache... 1126 images, 0 backgrounds, 0 corrupt: 100%|██████████| 1126/1126 [00:00<?, ?it/s]


WARNING ⚠️ cache='ram' may produce non-deterministic training results. Consider cache='disk' as a deterministic alternative if your disk space allows.


train: Caching images (2.9GB RAM): 100%|██████████| 1126/1126 [00:21<00:00, 51.32it/s]

albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))


val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1118.7±819.4 MB/s, size: 496.9 KB)


val: Scanning /content/dataset/labels/val... 282 images, 0 backgrounds, 0 corrupt: 100%|██████████| 282/282 [00:00<00:00, 1703.74it/s]

val: New cache created: /content/dataset/labels/val.cache


WARNING ⚠️ cache='ram' may produce non-deterministic training results. Consider cache='disk' as a deterministic alternative if your disk space allows.


val: Caching images (0.7GB RAM): 100%|██████████| 282/282 [00:08<00:00, 32.66it/s]


Plotting labels to drive/MyDrive/ml/var-ai/models/yolo11s-sports-ball-img1280-freeze10-1408/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.005' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.002, momentum=0.9) with parameter groups 81 weight(decay=0.0), 88 weight(decay=0.0005078125), 87 bias(decay=0.0)
Image sizes 1280 train, 1280 val
Using 2 dataloader workers
Logging results to drive/MyDrive/ml/var-ai/models/yolo11s-sports-ball-img1280-freeze10-1408
Starting training for 4.5 hours...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      1/100      7.32G     0.8777      2.386     0.9367         16       1280: 100%|██████████| 87/87 [01:18<00:00,  1.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:07<00:00,  1.51it/s]

                   all        282        282      0.914      0.945      0.951      0.746



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      2/187      7.32G     0.8672     0.6959     0.9404         11       1280: 100%|██████████| 87/87 [01:14<00:00,  1.17it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:05<00:00,  1.93it/s]

                   all        282        282      0.947      0.949      0.973      0.748



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      3/194      7.35G     0.8181     0.6167     0.9215         18       1280: 100%|██████████| 87/87 [01:12<00:00,  1.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:05<00:00,  2.15it/s]

                   all        282        282      0.982      0.943      0.975      0.769



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      4/198      7.33G     0.7825     0.5655     0.9207         11       1280: 100%|██████████| 87/87 [01:07<00:00,  1.30it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:05<00:00,  1.96it/s]

                   all        282        282      0.954      0.962      0.966      0.772



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      5/203      7.33G     0.7742     0.5096     0.9147         12       1280: 100%|██████████| 87/87 [01:08<00:00,  1.28it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:06<00:00,  1.68it/s]

                   all        282        282      0.981      0.961      0.985      0.769



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      6/205      7.36G     0.7754     0.4867     0.9215         11       1280: 100%|██████████| 87/87 [01:09<00:00,  1.26it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:05<00:00,  1.90it/s]

                   all        282        282      0.974      0.972      0.987      0.795



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      7/207      7.32G     0.7511     0.4707     0.9052         12       1280: 100%|██████████| 87/87 [01:09<00:00,  1.26it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:05<00:00,  2.10it/s]

                   all        282        282      0.993      0.952      0.985      0.816



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      8/208      7.32G     0.7276     0.4623     0.9047         11       1280: 100%|██████████| 87/87 [01:09<00:00,  1.25it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:06<00:00,  1.82it/s]

                   all        282        282      0.978      0.975      0.991      0.801



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      9/209      7.35G     0.7099     0.4452     0.8904         20       1280: 100%|██████████| 87/87 [01:07<00:00,  1.30it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:06<00:00,  1.78it/s]

                   all        282        282      0.997      0.972      0.993      0.822



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     10/210      7.32G     0.7115     0.4374     0.8875         12       1280: 100%|██████████| 87/87 [01:08<00:00,  1.28it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:05<00:00,  1.98it/s]

                   all        282        282      0.978      0.968      0.986      0.791



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     11/211      7.32G     0.7068     0.4326     0.8912         14       1280: 100%|██████████| 87/87 [01:08<00:00,  1.27it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:04<00:00,  2.22it/s]

                   all        282        282      0.979      0.977      0.994      0.824



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     12/212      7.35G     0.6984     0.4185     0.8856         13       1280: 100%|██████████| 87/87 [01:06<00:00,  1.31it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:04<00:00,  2.22it/s]

                   all        282        282      0.982      0.988      0.995      0.833



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     13/213      7.32G      0.682     0.4109     0.8825         12       1280: 100%|██████████| 87/87 [01:07<00:00,  1.30it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:05<00:00,  1.88it/s]

                   all        282        282      0.989      0.993      0.995      0.842



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     14/213      7.33G     0.6834     0.4194     0.8798         18       1280: 100%|██████████| 87/87 [01:06<00:00,  1.31it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:06<00:00,  1.71it/s]

                   all        282        282       0.98      0.979      0.993      0.851



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     15/214      7.36G     0.6641     0.4064     0.8815         15       1280: 100%|██████████| 87/87 [01:08<00:00,  1.27it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:05<00:00,  2.14it/s]

                   all        282        282      0.968      0.993      0.994      0.837



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     16/214      7.39G       0.66     0.3932     0.8735         18       1280: 100%|██████████| 87/87 [01:06<00:00,  1.30it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:05<00:00,  2.20it/s]

                   all        282        282      0.996      0.989      0.995       0.84



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     17/215      7.34G     0.6668     0.3973     0.8793         20       1280: 100%|██████████| 87/87 [01:07<00:00,  1.29it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:05<00:00,  2.18it/s]

                   all        282        282      0.995      0.982      0.994      0.834



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     18/215      7.34G     0.6403     0.3762     0.8734         14       1280: 100%|██████████| 87/87 [01:09<00:00,  1.25it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:06<00:00,  1.69it/s]

                   all        282        282      0.985      0.989      0.995       0.85



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     19/215      7.35G     0.6418      0.381     0.8717         20       1280: 100%|██████████| 87/87 [01:10<00:00,  1.24it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:05<00:00,  2.06it/s]

                   all        282        282      0.995      0.979      0.995      0.842



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     20/215      7.34G     0.6262     0.3658     0.8576         14       1280: 100%|██████████| 87/87 [01:07<00:00,  1.29it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:04<00:00,  2.23it/s]

                   all        282        282      0.981      0.986      0.994      0.831



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     21/215      7.34G      0.638     0.3807     0.8641         17       1280: 100%|██████████| 87/87 [01:06<00:00,  1.30it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:05<00:00,  2.14it/s]

                   all        282        282      0.982      0.988      0.994      0.846



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     22/216      7.36G     0.6081     0.3596     0.8517         16       1280: 100%|██████████| 87/87 [01:07<00:00,  1.28it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:05<00:00,  2.08it/s]

                   all        282        282      0.995      0.986      0.994      0.848



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     23/216      7.33G     0.6079     0.3579     0.8614         20       1280: 100%|██████████| 87/87 [01:07<00:00,  1.29it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:05<00:00,  1.86it/s]

                   all        282        282      0.989      0.987      0.994      0.841



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     24/216      7.33G     0.6245     0.3623      0.866         10       1280: 100%|██████████| 87/87 [01:05<00:00,  1.32it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:06<00:00,  1.81it/s]

                   all        282        282      0.988      0.986      0.994      0.849
EarlyStopping: Training stopped early as no improvement observed in last 10 epochs. Best results observed at epoch 14, best model saved as best.pt.
To update EarlyStopping(patience=10) pass a new patience value, i.e. `patience=300` or use `patience=0` to disable EarlyStopping.



24 epochs completed in 0.501 hours.
Optimizer stripped from drive/MyDrive/ml/var-ai/models/yolo11s-sports-ball-img1280-freeze10-1408/weights/last.pt, 19.2MB
Optimizer stripped from drive/MyDrive/ml/var-ai/models/yolo11s-sports-ball-img1280-freeze10-1408/weights/best.pt, 19.2MB

Validating drive/MyDrive/ml/var-ai/models/yolo11s-sports-ball-img1280-freeze10-1408/weights/best.pt...
Ultralytics 8.3.141 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
YOLO11s summary (fused): 100 layers, 9,413,187 parameters, 0 gradients, 21.3 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  27%|██▋       | 3/11 [00:02<00:06,  1.27it/s]